added from_base_string to int module

massivefermion opened the pull request.

massivefermion commented:
Sorry, this PR has a problem when compiled to javascript.
In the test I added, I required that:
```gleam
 "AG"
  |> int.from_base_string(16)
  |> should.equal(Error(Nil))
```
This works when the target is erlang, but when the target is javascript, I'm using `parseInt` and I'm returning error whenever the result of `parseInt` is `NaN`. But I forgot that `parseInt` does not return `NaN` if it can turn the string into a number by ignoring problematic parts. So we get `parseInt("AG", 16) == 10`.
It seems that to get the desired behavior, we'll need to manually check for invalid digits in the javascript side. Is this the only way?

lpil commented:
We will need to ensure that the behaviour is exactly the same in Erlang as in JavaScript so we may need to add some additional checking.

The other parsing function in this module is called parse, what could we name this function to make it more like that?

massivefermion commented:
I added a check for invalid digits, and now it errors out, which is the same as the erlang side.
About the name, maybe something like `parse_in_base` or `parse_with_base`.

inoas commented:
@massivefermion for the tests to pass you have to install gleam nightly:

```
asdf install gleam nightly
asdf local gleam nightly
gleam format
```

... commit and push.

p.s.: The formatter on this issue in map.gleam is flagged as buggy but that's another story.

massivefermion commented:
I renamed it to `base_parse` and added the documentation. Although I still think `parse_in_base` is a better name for grammatical reasons but It loses to `base_parse` if you consider brevity. So I leave it to you to pick a side in that trade-off.t

massivefermion commented:
👀

lpil reviewed (approved):
Thank you. 💜 
I've rebased this onto main
